### ISSUE TITLE
feat: add database support for envd-server

### ISSUE THREAD

**Issue by VoVAllen** [tensorchord/envd-server#32]
## Description

We can follow bytebase's practice that embeds a pg in the server binary. https://github.com/bytebase/bytebase/blob/740afc6286cd855fecc3cc54875583c6b650a41a/docs/design/source-code-tour.snb.md#bundling-postgresql

---

**Message from the maintainers**:

Love this enhancement proposal? Give it a 👍. We prioritise the proposals with the most 👍.

**Comment by gaocegege** [comment]
I think sqlite works in our case.

/cc @dragonly

**Comment by dragonly** [comment]
Thanks for referring to Bytebase! @VoVAllen @gaocegege 

# TL;DR

If you have tens of tables, migrate schema frequently, and plan to use some advanced features of PostgreSQL (e.g., time series data), then PostgreSQL is a proper choice. Otherwise, it’s a bit overkilling, and SQLite may be a more flexible choice.

# Longer version

I can provide some facts for you to decide. Bytebase started with SQLite and then migrated to PostgreSQL about eight months ago.

IMO, the main reason behind this is that we use the metadata database quite a lot. For example, we have **39 tables** now: https://github.com/bytebase/bytebase/blob/main/store/migration/prod/LATEST.sql.

Another reason is that we migrate schema **a lot**. Here is Bytebase’s [Schema Update Guide](https://github.com/bytebase/bytebase/blob/main/docs/schema-update-guide.md). In short, each release will have some [migration files](https://github.com/bytebase/bytebase/tree/main/store/migration/prod). When the new Bytebase binary starts, it’ll execute the migration on an earlier version of the meta database to upgrade the schema to the latest version.

Full story here: https://www.bytebase.com/blog/database-migration-sqlite-to-postgresql.

**Comment by gaocegege** [comment]
Thanks for your reply! @dragonly 

Since the envd-server is deployed on Kubernetes natively, I do not see the need of embedding pgsql into the bin. We can deploy pgsql as a separate deployment.

**Comment by VoVAllen** [comment]
ORM framework:
- https://gorm.io/gen/
- https://github.com/ent/ent
- https://gorm.io/

GORM is not type safe. Ent is GraphQL centric, might be too much for us. Seems gen is the good option

**Comment by VoVAllen** [comment]
Did some research on ORM in golang. Another mentioned option is sqlc. However many people said go doesn't need ORM at all.

Reference:
- https://www.reddit.com/r/golang/comments/umkgk3/gorm_is_a_bad_idea/
- https://www.reddit.com/r/golang/comments/t08oox/are_orms_considered_an_antipattern_in_go/?utm_medium=android_app&utm_source=share

**Comment by xieydd** [comment]
Whether or not to use the ORM framework depends on whether you are familiar with sql and whether your entities queries are complex.

**Comment by dragonly** [comment]
IMHO, working with SQL is easy utilizing `database/sql`.

Examples: https://github.com/bytebase/bytebase/tree/main/store

**Comment by VoVAllen** [comment]
Thanks suggestion form @dragonly and @xieydd . I finally decided to go with sqlc, which is SQL-centric utility library instead of a real ORM framework to reduce potential complexity.
